File: quanestimation/AsymptoticBound/Holevo.py
import numpy as np
import scipy as sp
import cvxpy as cp
from quanestimation.Common.Common import suN_generator
from quanestimation.AsymptoticBound.CramerRao import QFIM


def HCRB(rho, drho, W, eps=1e-8):
    """
    Calculation of the Holevo Cramer-Rao bound (HCRB) via the semidefinite program (SDP).

    Parameters
    ----------
    > **rho:** `matrix`
        -- Density matrix.

    > **drho:** `list`
        -- Derivatives of the density matrix on the unknown parameters to be 
        estimated. For example, drho[0] is the derivative vector on the first 
        parameter.

    > **W:** `matrix`
        -- Weight matrix.

    > **eps:** `float`
        -- Machine epsilon.

    Returns
    ----------
    **HCRB:** `float`
        -- The value of Holevo Cramer-Rao bound.
    """

    if type(drho) != list:
        raise TypeError("Please make sure drho is a list!")

    if len(drho) == 1:
        print(
            "In single parameter scenario, HCRB is equivalent to QFI. This function will return the value of QFI"
        )
        f = QFIM(rho, drho, eps=eps)
        return f
    else:
        dim = len(rho)
        num = dim * dim
        para_num = len(drho)

        Lambda = [np.identity(dim)] + suN_generator(dim)
        Lambda = Lambda / np.sqrt(2)

        vec_drho = [[] for i in range(para_num)]
        for pi in range(para_num):
            vec_drho[pi] = np.array(
                [
                    np.real(np.trace(np.dot(drho[pi], Lambda[i])))
                    for i in range(len(Lambda))
                ]
            )

        S = np.zeros((num, num), dtype=np.complex128)
        for a in range(num):
            for b in range(num):
                S[a][b] = np.trace(np.dot(Lambda[a], np.dot(Lambda[b], rho)))

        accu = len(str(int(1 / eps))) - 1
        lu, d, perm = sp.linalg.ldl(S.round(accu))
        R = np.dot(lu, sp.linalg.sqrtm(d)).conj().T
        # ============optimization variables================
        V = cp.Variable((para_num, para_num))
        X = cp.Variable((num, para_num))
        # ================add constraints===================
        constraints = [cp.bmat([[V, X.T @ R.conj().T], [R @ X, np.identity(num)]]) >> 0]
        for i in range(para_num):
            for j in range(para_num):
                if i == j:
                    constraints += [X[:, i].T @ vec_drho[j] == 1]
                else:
                    constraints += [X[:, i].T @ vec_drho[j] == 0]

        prob = cp.Problem(cp.Minimize(cp.trace(W @ V)), constraints)
        prob.solve()

        return prob.value

File: quanestimation/JuliaSrc/ObjectiveFunc/AsymptoticBound/Holevo.jl
function decomposition(A)
    C = bunchkaufman(A; check = false)
    R = sqrt(Array(C.D)) * C.U'C.P
    return R
end

"""

    HCRB(ρ::AbstractMatrix, dρ::AbstractVector, C::AbstractMatrix; eps=GLOBAL_EPS)

Caltulate the Holevo Cramer-Rao bound (HCRB) via the semidefinite program (SDP).
- `ρ`: Density matrix.
- `dρ`: Derivatives of the density matrix on the unknown parameters to be estimated. For example, drho[0] is the derivative vector on the first parameter.
- `W`: Weight matrix.
- `eps`: Machine epsilon.
"""
function HCRB(
    ρ::AbstractMatrix,
    dρ::AbstractVector,
    C::AbstractMatrix;
    eps=GLOBAL_EPS,
) 
    if length(dρ) == 1
        println(
            "In the single-parameter scenario, the HCRB is equivalent to the QFI. This function will return the value of the QFI.",
        )
        f = QFIM_SLD(ρ, dρ[1]; eps=eps)
        return f
    elseif rank(C) == 1
        println(
            "For rank-one wight matrix, the HCRB is equivalent to QFIM. This function will return the value of Tr(WF^{-1})."
        )
        F = QFIM_SLD(ρ, dρ; eps=eps)
        return tr(C*pinv(F))
    else
        Holevo_bound(ρ, dρ, C; eps = eps)
    end
end

function Holevo_bound(
    ρ::AbstractMatrix,
    dρ::AbstractVector,
    C::AbstractMatrix;
    eps = GLOBAL_EPS,
) 

    dim = size(ρ)[1]
    num = dim * dim
    para_num = length(dρ)
    suN = suN_generator(dim) / sqrt(2)
    Lambda = [Matrix{ComplexF64}(I, dim, dim) / sqrt(2)]
    append!(Lambda, [suN[i] for i = 1:length(suN)])
    vec_∂ρ = [[0.0 for i = 1:num] for j = 1:para_num]

    for pa = 1:para_num
        for ra = 2:num
            vec_∂ρ[pa][ra] = (dρ[pa] * Lambda[ra]) |> tr |> real
        end
    end
    S = zeros(ComplexF64, num, num)
    for a = 1:num
        for b = 1:num
            S[a, b] = (Lambda[a] * Lambda[b] * ρ) |> tr
        end
    end

    accu = length(string(Int(ceil(1 / eps)))) - 1
    R = decomposition(round.(digits = accu, S))

    #=========optimization variables===========#
    V = Variable(para_num, para_num)
    X = Variable(num, para_num)
    #============add constraints===============#
    constraints = [[V X'*R'; R*X Matrix{Float64}(I, num, num)] ⪰ 0]
    for i = 1:para_num
        for j = 1:para_num
            if i == j
                constraints += [X[:, i]' * vec_∂ρ[j] == 1]
            else
                constraints += [X[:, i]' * vec_∂ρ[j] == 0]
            end
        end
    end
    problem = minimize(tr(C * V), constraints)
    Convex.solve!(problem, SCS.Optimizer, silent_solver=true)
    return evaluate(tr(C * V))
end

function Holevo_bound_obj(
    ρ::AbstractMatrix,
    dρ::AbstractVector,
    C::AbstractMatrix;
    eps = GLOBAL_EPS,
)
    return Holevo_bound(ρ, dρ, C; eps = eps)[1]
end
